File: windows_desktop/speech/banglish_transliteration.py
"""
Bongo Type - Banglish Transliteration Engine
Converts Bengali spoken text to authentic Romanized Latin script (Banglish),
preserving phonetics without transforming into Bengali script.
"""

import re

COMMON_WORDS = {
    "আমি": "Ami",
    "আমার": "amar",
    "আমরা": "amra",
    "তুমি": "Tumi",
    "তোমার": "tomar",
    "তোমরা": "tomra",
    "সে": "she",
    "তার": "tar",
    "তারা": "tara",
    "সাথে": "sathe",
    "কথা": "kotha",
    "বলছি": "bolchi",
    "বলছিলেন": "bolchilen",
    "বলবে": "bolbe",
    "এখন": "ekhon",
    "কেমন": "kemon",
    "আছো": "acho",
    "আছেন": "achen",
    "ধন্যবাদ": "dhonnobad",
    "বাংলাদেশ": "Bangladesh",
    "ভালো": "bhalo",
    "ভাল": "bhalo",
    "কি": "ki",
    "কী": "ki",
    "কোথায়": "kothay",
    "কোথা": "kotha",
    "কেন": "keno",
    "হ্যাঁ": "haan",
    "না": "na",
    "খুব": "khub",
    "সুন্দর": "shundor",
    "দিন": "din",
    "রাত": "raat",
    "কাজ": "kaaj",
    "করছি": "korchi",
    "করবো": "korbo",
    "ভালোবাসি": "bhalobashi",
    "বন্ধু": "bondhu",
    "ভাই": "bhai",
    "একটু": "ektu",
    "অনেক": "onek",
    "আজকে": "aajke",
    "কালকে": "kaalke",
    "ঠিক": "thik",
    "আছে": "ache"
}

VOWELS = {
    'অ': 'o', 'আ': 'a', 'ই': 'i', 'ঈ': 'ee', 'উ': 'u',
    'ঊ': 'oo', 'ঋ': 'ri', 'এ': 'e', 'ঐ': 'oi', 'ও': 'o', 'ঔ': 'ou'
}

VOWEL_MARKS = {
    'া': 'a', 'ি': 'i', 'ী': 'ee', 'ু': 'u', 'ূ': 'oo',
    'ৃ': 'ri', 'ে': 'e', 'ৈ': 'oi', 'ো': 'o', 'ৌ': 'ou'
}

CONSONANTS = {
    'ক': 'k', 'খ': 'kh', 'গ': 'g', 'ঘ': 'gh', 'ঙ': 'ng',
    'চ': 'ch', 'ছ': 'chh', 'জ': 'j', 'ঝ': 'jh', 'ঞ': 'n',
    'ট': 't', 'ঠ': 'th', 'ড': 'd', 'ঢ': 'dh', 'ণ': 'n',
    'ত': 't', 'থ': 'th', 'দ': 'd', 'ধ': 'dh', 'ন': 'n',
    'প': 'p', 'ফ': 'f', 'ব': 'b', 'ভ': 'bh', 'ম': 'm',
    'য': 'j', 'র': 'r', 'ল': 'l', 'শ': 'sh', 'ষ': 'sh',
    'স': 's', 'হ': 'h', 'ড়': 'r', 'ঢ়': 'rh', 'য়': 'y',
    'ৎ': 't', 'ং': 'ng', 'ঃ': 'h', 'ঁ': 'n'
}


def to_banglish(text: str) -> str:
    """Converts Bengali speech output to natural Banglish Latin script."""
    text = text.strip()
    if not text:
        return ""

    # Check if already primarily Latin script
    has_bengali = any('\u0980' <= c <= '\u09FF' for c in text)
    if not has_bengali:
        return text

    words = re.split(r'\s+', text)
    converted_words = []

    for idx, word in enumerate(words):
        clean_word = re.sub(r'[।.,!?:;\"\'()_]+', '', word)
        trailing_punct = ""
        for c in reversed(word):
            if c in "।.,!?:;\"'()":
                trailing_punct = ('.' if c == '।' else c) + trailing_punct
            else:
                break

        if clean_word in COMMON_WORDS:
            res = COMMON_WORDS[clean_word]
            if idx == 0:
                res = res.capitalize()
            else:
                res = res.lower()
        else:
            res = phonetic_transliterate(clean_word)

        converted_words.append(res + trailing_punct)

    result = " ".join(converted_words)
    if result:
        result = result[0].upper() + result[1:]
    return result


def phonetic_transliterate(word: str) -> str:
    """Fallback phonetic converter using rule-based mapping."""
    res = []
    i = 0
    length = len(word)

    while i < length:
        ch = word[i]
        if ch in VOWELS:
            res.append(VOWELS[ch])
        elif ch in VOWEL_MARKS:
            res.append(VOWEL_MARKS[ch])
        elif ch in CONSONANTS:
            res.append(CONSONANTS[ch])
            if i + 1 < length:
                next_ch = word[i + 1]
                if next_ch == '্':  # Hasanta / conjunct
                    i += 1
                elif next_ch not in VOWEL_MARKS and next_ch in CONSONANTS:
                    if i + 2 < length and word[i + 2] not in VOWEL_MARKS:
                        res.append('o')
        else:
            res.append(ch)
        i += 1

    return "".join(res)

File: app/src/main/java/com/example/speech/BanglishTransliteration.kt
package com.example.speech

object BanglishTransliteration {

    private val commonWordMap = mapOf(
        "আমি" to "Ami",
        "আমার" to "amar",
        "আমায়" to "amay",
        "আমরা" to "amra",
        "আমাদের" to "amader",
        "তুমি" to "Tumi",
        "তোমার" to "tomar",
        "তোমায়" to "tomay",
        "তোমরা" to "tomra",
        "তোমাদের" to "tomader",
        "আপনি" to "Apni",
        "আপনার" to "apnar",
        "আপনাকে" to "apnake",
        "সে" to "she",
        "তার" to "tar",
        "তাকে" to "take",
        "তারা" to "tara",
        "তাদের" to "tader",
        "তিনি" to "tini",
        "তাঁর" to "tar",
        "সাথে" to "sathe",
        "কথা" to "kotha",
        "বলছি" to "bolchi",
        "বলছিলে" to "bolchhile",
        "বলবে" to "bolbe",
        "বলেছি" to "bolechi",
        "বলুন" to "bolun",
        "এখন" to "ekhon",
        "কেমন" to "kemon",
        "আছো" to "acho",
        "আছেন" to "achen",
        "আছি" to "achi",
        "ধন্যবাদ" to "dhonnobad",
        "বাংলাদেশ" to "Bangladesh",
        "ভালো" to "bhalo",
        "ভাল" to "bhalo",
        "কি" to "ki",
        "কী" to "ki",
        "কোথায়" to "kothay",
        "কোথা" to "kotha",
        "কেন" to "keno",
        "হ্যাঁ" to "haan",
        "না" to "na",
        "খুব" to "khub",
        "সুন্দর" to "shundor",
        "দিন" to "din",
        "রাত" to "raat",
        "কাজ" to "kaaj",
        "করছি" to "korchi",
        "করবো" to "korbo",
        "করছেন" to "korchen",
        "করবেন" to "korben",
        "করো" to "koro",
        "যাবো" to "jabo",
        "যাচ্ছি" to "jachhi",
        "যাবেন" to "jaben",
        "আসছি" to "ashchi",
        "আসবেন" to "ashben",
        "ভালোবাসি" to "bhalobashi",
        "মানুষ" to "manush",
        "বন্ধু" to "bondhu",
        "ভাই" to "bhai",
        "কে" to "ke",
        "কাকে" to "kake",
        "কখন" to "kokhon",
        "একটু" to "ektu",
        "অনেক" to "onek",
        "সবাই" to "shobai",
        "আজকে" to "aajke",
        "কালকে" to "kaalke",
        "ঠিক" to "thik",
        "আছে" to "ache",
        "যাও" to "jao",
        "এসো" to "esho",
        "খবর" to "khobor",
        "মেসেজ" to "message",
        "টাইপ" to "type"
    )

    private val vowels = mapOf(
        'অ' to "o",
        'আ' to "a",
        'ই' to "i",
        'ঈ' to "ee",
        'উ' to "u",
        'ঊ' to "oo",
        'ঋ' to "ri",
        'এ' to "e",
        'ঐ' to "oi",
        'ও' to "o",
        'ঔ' to "ou"
    )

    private val vowelMarks = mapOf(
        'া' to "a",
        'ি' to "i",
        'ী' to "ee",
        'ু' to "u",
        'ূ' to "oo",
        'ৃ' to "ri",
        'ে' to "e",
        'ৈ' to "oi",
        'ো' to "o",
        'ৌ' to "ou"
    )

    private val consonants = mapOf(
        'ক' to "k", 'খ' to "kh", 'গ' to "g", 'ঘ' to "gh", 'ঙ' to "ng",
        'চ' to "ch", 'ছ' to "chh", 'জ' to "j", 'ঝ' to "jh", 'ঞ' to "n",
        'ট' to "t", 'ঠ' to "th", 'ড' to "d", 'ঢ' to "dh", 'ণ' to "n",
        'ত' to "t", 'থ' to "th", 'দ' to "d", 'ধ' to "dh", 'ন' to "n",
        'প' to "p", 'ফ' to "f", 'ব' to "b", 'ভ' to "bh", 'ম' to "m",
        'য' to "j", 'র' to "r", 'ল' to "l", 'শ' to "sh", 'ষ' to "sh",
        'স' to "s", 'হ' to "h", 'ড়' to "r", 'ঢ়' to "rh", 'য়' to "y",
        'ৎ' to "t", 'ং' to "ng", 'ঃ' to "h", 'ঁ' to "n"
    )

    fun toBanglish(bengaliText: String): String {
        val trimmed = bengaliText.trim()
        if (trimmed.isEmpty()) return ""

        // Check if input is already in Latin script
        val hasBengaliChars = trimmed.any { it in '\u0980'..'\u09FF' }
        if (!hasBengaliChars) {
            return trimmed
        }

        val words = trimmed.split(Regex("\\s+"))
        val convertedWords = words.mapIndexed { index, word ->
            // Strip punctuation for matching
            val cleanWord = word.replace(Regex("[।.,!?:;\"'()_]+"), "")
            val punctuationSuffix = word.takeLastWhile { it in "।.,!?:;\"'()" }.replace("।", ".")

            val matched = commonWordMap[cleanWord]
            val result = if (matched != null) {
                if (index == 0) matched.replaceFirstChar { it.uppercase() } else matched.lowercase()
            } else {
                phoneticConvertWord(cleanWord)
            }
            result + punctuationSuffix
        }

        val finalString = convertedWords.joinToString(" ")
        return finalString.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
    }

    private fun phoneticConvertWord(word: String): String {
        if (word.isEmpty()) return ""
        val sb = StringBuilder()
        var i = 0
        val len = word.length

        while (i < len) {
            val ch = word[i]

            when {
                vowels.containsKey(ch) -> {
                    sb.append(vowels[ch])
                }
                vowelMarks.containsKey(ch) -> {
                    sb.append(vowelMarks[ch])
                }
                consonants.containsKey(ch) -> {
                    val cons = consonants[ch] ?: ""
                    sb.append(cons)

                    // Lookahead: check if next character is vowel mark, hasanta, or next consonant
                    if (i + 1 < len) {
                        val nextCh = word[i + 1]
                        if (nextCh == '্') {
                            // Hasanta (conjunct connector)
                            i += 1 // skip hasanta
                        } else if (!vowelMarks.containsKey(nextCh) && consonants.containsKey(nextCh)) {
                            // Inherent 'o' or 'a' between consonants if not ending
                            if (i + 2 < len && !vowelMarks.containsKey(word[i + 2])) {
                                sb.append("o")
                            }
                        }
                    }
                }
                else -> {
                    sb.append(ch)
                }
            }
            i++
        }
        return sb.toString()
    }
}
